ably/auth.go

```diff
@@ -266,8 +271,13 @@ func (a *Auth) Authorize(ctx context.Context, params *TokenParams, setOpts ...Au
 		opts = applyAuthOptionsWithDefaults(setOpts...)
 	}
 	a.mtx.Lock()
-	defer a.mtx.Unlock()
-	return a.authorize(ctx, params, opts, true)
+	token, err := a.authorize(ctx, params, opts, true)
+	a.mtx.Unlock()
```

Comment:
It feels wrong to hold a mutex for the duration of a remote request, what exactly is this protecting? Does this block anything that needs access to the current auth details whilst the request is in flight?

Reply:
Auth remembers (caches) the last token, among other auth-related stuff. That's what this mutex protects.

I'm not sure if anything relies on this lock covering the whole request, but it makes some intuitive sense to me because otherwise several parallel authorization requests could be launched that would then race to set the current token, and then callers wouldn't agree on which the current token is, etc. It probably could be changed to a RWMutex, though.

(Note that this isn't new in this PR nor related to RTC8a, so, if needed, can be analyzed and fixed separately.)

Reply:
> Auth remembers (caches) the last token, among other auth-related stuff. That's what this mutex protects.

Ok, so I do think a mutex lock should always wrap _just_ the memory addresses it's protecting in an obvious way, since in my experience, the further away the wrapping gets, the more likely we are to encounter deadlocks.

> several parallel authorization requests could be launched that would then race to set the current token, and then callers wouldn't agree on which the current token is, etc. It probably could be changed to a RWMutex, though.

Ok I understand, in that case then I don't think a mutex is the correct primitive to use, I would suggest a [`singleflight.Group`](https://pkg.go.dev/golang.org/x/sync/singleflight#Group.Do).

> Note that this isn't new in this PR nor related to RTC8a, so, if needed, can be analyzed and fixed separately.

SGTM 👍

Reply:
Singleflight doesn't really fit here because the params passed to Authorize could be different, so calls wouldn't be expected to be idempotent. You shouldn't just return the return value of another concurrent call.

I've linked to this conversation in a relevant issue: https://github.com/ably/ably-go/issues/184